File: pregnancy_diary.php
# pregnancy_diary.php
#!/usr/bin/env php
<?php

define('DATA_FILE', 'pregnancy_diary.json');

class Note {
    public $id;
    public $content;
    public $category;
    public $date;
    public $weight;
    public $mood;
    public $created_at;

    function __construct($content, $category = 'general', $date = null, $weight = null, $mood = null) {
        $this->id = substr(bin2hex(random_bytes(4)), 0, 8);
        $this->content = $content;
        $this->category = $category;
        $this->date = $date ?: date('Y-m-d');
        $this->weight = $weight;
        $this->mood = $mood;
        $this->created_at = date('c');
    }

    function toArray() {
        return [
            'id' => $this->id,
            'content' => $this->content,
            'category' => $this->category,
            'date' => $this->date,
            'weight' => $this->weight,
            'mood' => $this->mood,
            'created_at' => $this->created_at
        ];
    }

    static function fromArray($data) {
        $n = new self($data['content'], $data['category'], $data['date'], $data['weight'], $data['mood']);
        $n->id = $data['id'];
        $n->created_at = $data['created_at'];
        return $n;
    }
}

class Diary {
    private $notes = [];

    function __construct() {
        $this->load();
    }

    function load() {
        if (file_exists(DATA_FILE)) {
            $data = json_decode(file_get_contents(DATA_FILE), true);
            $this->notes = array_map(function($d) { return Note::fromArray($d); }, $data);
        }
    }

    function save() {
        $data = array_map(function($n) { return $n->toArray(); }, $this->notes);
        file_put_contents(DATA_FILE, json_encode($data, JSON_PRETTY_PRINT));
    }

    function add($content, $category = 'general', $date = null, $weight = null, $mood = null) {
        $n = new Note($content, $category, $date, $weight, $mood);
        $this->notes[] = $n;
        $this->save();
        echo "✅ Note added (ID: {$n->id})\n";
    }

    function list($category = null, $since = null) {
        $notes = $this->notes;
        if ($category) {
            $notes = array_filter($notes, function($n) use ($category) {
                return $n->category == $category;
            });
        }
        if ($since) {
            $notes = array_filter($notes, function($n) use ($since) {
                return $n->date >= $since;
            });
        }
        if (empty($notes)) {
            echo "No matching notes.\n";
            return;
        }
        echo "\n📋 Pregnancy Diary (" . count($notes) . " entries):\n\n";
        foreach ($notes as $n) {
            $weightStr = $n->weight ? " – {$n->weight} kg" : '';
            $moodStr = $n->mood ? " ⭐" . str_repeat('⭐', $n->mood-1) : '';
            echo "{$n->date}: [{$n->category}] {$n->content}{$weightStr}{$moodStr}\n";
        }
    }

    function stats() {
        if (empty($this->notes)) {
            echo "No entries yet.\n";
            return;
        }
        $cats = [];
        $moods = [];
        $weights = [];
        foreach ($this->notes as $n) {
            $cats[$n->category] = ($cats[$n->category] ?? 0) + 1;
            if ($n->mood) $moods[] = $n->mood;
            if ($n->weight) $weights[] = $n->weight;
        }
        echo "\n📊 Statistics:\n";
        echo "  Total entries: " . count($this->notes) . "\n";
        foreach ($cats as $cat => $count) {
            echo "  " . ucfirst($cat) . " entries: $count\n";
        }
        if (!empty($moods)) {
            $avg = array_sum($moods) / count($moods);
            echo "  Average mood: " . round($avg, 1) . "/5\n";
        }
        if (!empty($weights)) {
            echo "  Latest weight: " . end($weights) . " kg\n";
        }
    }

    function search($term) {
        $results = array_filter($this->notes, function($n) use ($term) {
            return stripos($n->content, $term) !== false;
        });
        if (empty($results)) {
            echo "No matches.\n";
            return;
        }
        echo "\n🔍 Found " . count($results) . " note(s):\n";
        foreach ($results as $n) {
            echo "{$n->date}: [{$n->category}] {$n->content}\n";
        }
    }

    function exportHTML($filename = 'diary.html') {
        $html = '<!DOCTYPE html>
<html><head><title>Pregnancy Diary</title>
<style>body{font-family:sans-serif;margin:30px;background:#f5f5f5;}
.note{background:#fff;border-radius:8px;padding:15px;margin:10px 0;box-shadow:0 2px 4px rgba(0,0,0,0.1);}
.meta{color:#999;font-size:0.9em;}
</style></head><body>
<h1>🤰 Pregnancy Diary</h1>
';
        foreach (array_reverse($this->notes) as $n) {
            $html .= "<div class=\"note\">\n";
            $html .= "<div class=\"meta\">{$n->date} – " . ucfirst($n->category) . "</div>\n";
            $html .= "<div>{$n->content}</div>\n";
            if ($n->weight) $html .= "<div>Weight: {$n->weight} kg</div>\n";
            if ($n->mood) $html .= "<div>Mood: " . str_repeat('⭐', $n->mood) . "</div>\n";
            $html .= "</div>\n";
        }
        $html .= '</body></html>';
        file_put_contents($filename, $html);
        echo "✅ Exported to $filename\n";
    }
}

if ($argc < 2) {
    die("Usage: php pregnancy_diary.php <command> [options]\n");
}
$d = new Diary();
$cmd = $argv[1];

switch ($cmd) {
    case 'add':
        if ($argc < 3) die("add <content> [--category CAT] [--date DATE] [--weight KG] [--mood RATING]\n");
        $content = $argv[2];
        $category = 'general';
        $date = null;
        $weight = null;
        $mood = null;
        for ($i=3; $i<$argc; $i++) {
            if ($argv[$i] == '--category' && isset($argv[$i+1])) { $category = $argv[++$i]; }
            if ($argv[$i] == '--date' && isset($argv[$i+1])) { $date = $argv[++$i]; }
            if ($argv[$i] == '--weight' && isset($argv[$i+1])) { $weight = (float)$argv[++$i]; }
            if ($argv[$i] == '--mood' && isset($argv[$i+1])) { $mood = (int)$argv[++$i]; }
        }
        $d->add($content, $category, $date, $weight, $mood);
        break;

    case 'list':
        $category = null;
        $since = null;
        for ($i=2; $i<$argc; $i++) {
            if ($argv[$i] == '--category' && isset($argv[$i+1])) { $category = $argv[++$i]; }
            if ($argv[$i] == '--since' && isset($argv[$i+1])) { $since = $argv[++$i]; }
        }
        $d->list($category, $since);
        break;

    case 'stats':
        $d->stats();
        break;

    case 'search':
        if ($argc < 3) die("search <term>\n");
        $d->search($argv[2]);
        break;

    case 'export':
        $filename = 'diary.html';
        for ($i=2; $i<$argc; $i++) {
            if ($argv[$i] == '--filename' && isset($argv[$i+1])) { $filename = $argv[++$i]; }
        }
        $d->exportHTML($filename);
        break;

    default:
        echo "Unknown command. Use add, list, stats, search, export.\n";
}
?>

File: pregnancy_diary.rb
# pregnancy_diary.rb
#!/usr/bin/env ruby
require 'json'
require 'securerandom'
require 'date'

DATA_FILE = 'pregnancy_diary.json'

class Note
  attr_accessor :id, :content, :category, :date, :weight, :mood, :created_at

  def initialize(content, category = 'general', date = nil, weight = nil, mood = nil)
    @id = SecureRandom.hex(4)
    @content = content
    @category = category
    @date = date || Date.today.to_s
    @weight = weight
    @mood = mood
    @created_at = Time.now.iso8601
  end

  def to_hash
    {
      id: @id, content: @content, category: @category, date: @date,
      weight: @weight, mood: @mood, created_at: @created_at
    }
  end

  def self.from_hash(h)
    note = new(h['content'], h['category'], h['date'], h['weight'], h['mood'])
    note.id = h['id']
    note.created_at = h['created_at']
    note
  end
end

class Diary
  attr_reader :notes

  def initialize
    @notes = []
    load
  end

  def load
    if File.exist?(DATA_FILE)
      data = JSON.parse(File.read(DATA_FILE))
      @notes = data.map { |h| Note.from_hash(h) }
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@notes.map(&:to_hash)))
  end

  def add(content, category = 'general', date = nil, weight = nil, mood = nil)
    note = Note.new(content, category, date, weight, mood)
    @notes << note
    save
    puts "✅ Note added (ID: #{note.id})"
  end

  def list(category = nil, since = nil)
    notes = @notes
    notes = notes.select { |n| n.category == category } if category
    notes = notes.select { |n| n.date >= since } if since
    if notes.empty?
      puts "No matching notes."
      return
    end
    puts "\n📋 Pregnancy Diary (#{notes.size} entries):\n"
    notes.each do |n|
      weight_str = n.weight ? " – #{n.weight} kg" : ""
      mood_str = n.mood ? " ⭐#{'⭐' * (n.mood - 1)}" : ""
      puts "#{n.date}: [#{n.category}] #{n.content}#{weight_str}#{mood_str}"
    end
  end

  def stats
    if @notes.empty?
      puts "No entries yet."
      return
    end
    cats = Hash.new(0)
    moods = []
    weights = []
    @notes.each do |n|
      cats[n.category] += 1
      moods << n.mood if n.mood
      weights << n.weight if n.weight
    end
    puts "\n📊 Statistics:"
    puts "  Total entries: #{@notes.size}"
    cats.each { |cat, count| puts "  #{cat.capitalize} entries: #{count}" }
    unless moods.empty?
      avg = moods.sum.to_f / moods.size
      puts "  Average mood: #{avg.round(1)}/5"
    end
    puts "  Latest weight: #{weights.last} kg" unless weights.empty?
  end

  def search(term)
    results = @notes.select { |n| n.content.downcase.include?(term.downcase) }
    if results.empty?
      puts "No matches."
      return
    end
    puts "\n🔍 Found #{results.size} note(s):"
    results.each { |n| puts "#{n.date}: [#{n.category}] #{n.content}" }
  end

  def export_html(filename = 'diary.html')
    html = <<~HTML
    <!DOCTYPE html>
    <html><head><title>Pregnancy Diary</title>
    <style>body{font-family:sans-serif;margin:30px;background:#f5f5f5;}
    .note{background:#fff;border-radius:8px;padding:15px;margin:10px 0;box-shadow:0 2px 4px rgba(0,0,0,0.1);}
    .meta{color:#999;font-size:0.9em;}
    </style></head><body>
    <h1>🤰 Pregnancy Diary</h1>
    HTML
    @notes.reverse.each do |n|
      html += <<~NOTE
      <div class="note">
      <div class="meta">#{n.date} – #{n.category.capitalize}</div>
      <div>#{n.content}</div>
      NOTE
      html += "<div>Weight: #{n.weight} kg</div>" if n.weight
      html += "<div>Mood: #{'⭐' * n.mood}</div>" if n.mood
      html += "</div>"
    end
    html += "</body></html>"
    File.write(filename, html)
    puts "✅ Exported to #{filename}"
  end
end

if ARGV.empty?
  puts "Usage: pregnancy_diary.rb <command> [options]"
  exit
end

d = Diary.new
cmd = ARGV.shift

case cmd
when 'add'
  if ARGV.empty?
    puts "add <content> [--category CAT] [--date DATE] [--weight KG] [--mood RATING]"
    exit
  end
  content = ARGV.shift
  category = 'general'
  date = nil
  weight = nil
  mood = nil
  while ARGV.any?
    case ARGV[0]
    when '--category'
      ARGV.shift
      category = ARGV.shift || 'general'
    when '--date'
      ARGV.shift
      date = ARGV.shift
    when '--weight'
      ARGV.shift
      weight = ARGV.shift.to_f
    when '--mood'
      ARGV.shift
      mood = ARGV.shift.to_i
    else
      break
    end
  end
  d.add(content, category, date, weight, mood)

when 'list'
  category = nil
  since = nil
  if ARGV.include?('--category')
    idx = ARGV.index('--category')
    category = ARGV[idx+1] if idx
  end
  if ARGV.include?('--since')
    idx = ARGV.index('--since')
    since = ARGV[idx+1] if idx
  end
  d.list(category, since)

when 'stats'
  d.stats

when 'search'
  term = ARGV.shift
  if term.nil?
    puts "search <term>"
    exit
  end
  d.search(term)

when 'export'
  filename = 'diary.html'
  if ARGV.include?('--filename')
    idx = ARGV.index('--filename')
    filename = ARGV[idx+1] if idx
  end
  d.export_html(filename)

else
  puts "Unknown command. Use add, list, stats, search, export."
end
